### PR TITLE
Added RangeNotSatisfiableHttpException that uses when the requested range is not satisfiable

### PR DESCRIPTION
**Pull request by zalatov**
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | yes |
| Breaks BC? | no |
| Tests pass? | yes |
| Fixed issues |  |

It's need to catch this error.
For example, we can't need to log this error.
Now we can do `try...catch` or add it to exclude category.

### CODE REVIEW

File: framework/web/ContentRangeHttpException.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+class ContentRangeHttpException extends HttpException
+{
+    /**
+     * Constructor.
+     * @param integer $status HTTP status code, default is 416.
+     * @param string $message error message
+     * @param integer $code error code
+     * @param \Exception $previous The previous exception used for the exception chaining.
+     */
+    public function __construct($status = 416, $message = 'Requested range not satisfiable', $code = 0, \Exception $previous = null)
```
Comment (cebe) [inline framework/web/ContentRangeHttpException.php]: code should be hardcoded as it is the case with other Http exceptions: https://github.com/yiisoft/yii2/blob/af0119a910b9d6edd81d55d0b9395f27192edb2f/framework/web/NotFoundHttpException.php#L24

Comment (cebe) [inline framework/web/ContentRangeHttpException.php]: message should be `null` by default, default message is set by response if not present.

File: framework/web/ContentRangeHttpException.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+namespace yii\web;
+
+/**
+ * ContentRangeHttpException represents an exception caused by an improper request of the end-user.
+ * This exception uses when the requested range is not satisfiable.
+ *
+ * Throwing an ContentRangeHttpException like in the following example will result in the 416 page to be displayed.
+ *
+ * @author Залатов Александр <CaHbKa.Z@gmail.com>
+ */
```
Comment (cebe) [inline framework/web/ContentRangeHttpException.php]: `@since 2.0.11`

**Review by cebe — CHANGES_REQUESTED** [review]

File: framework/CHANGELOG.md [hunk 3]
```diff
@@ -58,6 +58,7 @@ Yii Framework 2 Change Log
 - Chg #10921: Inverts responsibility of database specific column schema builder classes (df2)
 - Chg #11071: `yii\helpers\BaseArrayHelper::isIn()` and `isTraversable()` since now throw `\yii\base\InvalidParamException` instead of `\InvalidArgumentException` (nukkumatti)
 - New #8920: Added `yii\mutex\PgsqlMutex` which implements mutex "lock" mechanism via PgSQL locks (nineinchnick, CSharpRU)
+- New: Added ContentRangeHttpException that uses when the requested range is not satisfiable (zalatov)
```
Comment (samdark) [inline framework/CHANGELOG.md]: Need `#12854` here.

File: framework/web/ContentRangeHttpException.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+class ContentRangeHttpException extends HttpException
```
Comment (samdark) [inline framework/web/ContentRangeHttpException.php]: How about renaming it to `InvalidContentRangeException`?

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/web/ContentRangeHttpException.php (hunk 1 above)
Comment (zalatov) [inline framework/web/ContentRangeHttpException.php]: fixed

File: framework/web/ContentRangeHttpException.php (hunk 2 above)
Comment (zalatov) [inline framework/web/ContentRangeHttpException.php]: fixed

File: framework/CHANGELOG.md (hunk 3 above)
Comment (zalatov) [inline framework/CHANGELOG.md]: added

I'm add a new section - 2.0.11

File: framework/CHANGELOG.md [hunk 5]
```diff
@@ -15,6 +15,7 @@ Yii Framework 2 Change Log
 - Bug #12824: Enabled usage of `yii\mutex\FileMutex` on Windows systems (davidsonalencar)
 - Enh #11037 yii.js and yii.validation.js should use Regexp.test instead of String.match (arogachev, nkovacs)
 - Bug #9796: Initialization of not existing `yii\grid\ActionColumn` default buttons (arogachev)
+- Enh #12854: Added `InvalidContentRangeHttpException` that uses when the requested range is not satisfiable (zalatov)
```
Comment (samdark) [inline framework/CHANGELOG.md]: The class name here in changelog and in the actual code is different.

**Review by samdark — CHANGES_REQUESTED** [review]

**Comment by zalatov** [comment]
fixed

**Comment by dynasource** [comment]
where is the unit test? Can you please add it?

File: framework/web/ContentRangeHttpException.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * ContentRangeHttpException represents an exception caused by an improper request of the end-user.
+ * This exception uses when the requested range is not satisfiable.
+ *
+ * Throwing an ContentRangeHttpException like in the following example will result in the 416 page to be displayed.
+ *
+ * @author Zalatov Alexander <CaHbKa.Z@gmail.com>
+ *
+ * @since 2.0.11
+ */
+class ContentRangeHttpException extends HttpException
```
Comment (rob006) [inline framework/web/ContentRangeHttpException.php]: Exception name is inconsistent - all other HttpExceptions uses [status code name](https://github.com/yiisoft/yii2/blob/master/framework/web/Response.php#L163) as exception name, so to this should be named `RangeNotSatisfiableHttpException`.

**Review by rob006 — CHANGES_REQUESTED** [review]

File: framework/web/ContentRangeHttpException.php (hunk 6 above)
Comment (samdark) [inline framework/web/ContentRangeHttpException.php]: Agree. Thanks for noticing that. @zalatov could you rename it?

**Comment by zalatov** [comment]
Sorry for so delay...

Fixed - renamed to RangeNotSatisfiableHttpException.

**Review by samdark — APPROVED** [review]

File: framework/CHANGELOG.md [hunk 7]
```diff
@@ -45,7 +45,7 @@ Yii Framework 2 Change Log
 - Enh #13035: Use ArrayHelper::getValue() in SluggableBehavior::getValue() (thyseus)
 - Enh #13020: Added `disabledListItemSubTagOptions` attribute for `yii\widgets\LinkPager` in order to customize the disabled list item sub tag element (nadar)
 - Enh: Added constants for specifying `yii\validators\CompareValidator::$type` (cebe)
-
+- Enh #12854: Added `RangeNotSatisfiableHttpException` that uses when the requested range is not satisfiable (zalatov)
```
Comment (dynasource) [inline framework/CHANGELOG.md]: ->  Added `RangeNotSatisfiableHttpException` to cover 416 file request exceptions

File: framework/web/RangeNotSatisfiableHttpException.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+ * @link http://www.yiiframework.com/
+ * @copyright Copyright (c) 2008 Yii Software LLC
+ * @license http://www.yiiframework.com/license/
+ */
+
+namespace yii\web;
+
+/**
+ * RangeNotSatisfiableHttpException represents an exception caused by an improper request of the end-user.
+ * This exception uses when the requested range is not satisfiable.
```
Comment (dynasource) [inline framework/web/RangeNotSatisfiableHttpException.php]: Please be more clarifying about this error. Its unclear that this is about files. Wikipedia states:
'The client has asked for a portion of the file (byte serving), but the server cannot supply that portion. For example, if the client asked for a part of the file that lies beyond the end of the file.'

File: framework/web/Response.php [hunk 9]
```diff
@@ -533,7 +533,7 @@ public function sendContentAsFile($content, $attachmentName, $options = [])
      *    This option is available since version 2.0.4.
      *
      * @return $this the response object itself
-     * @throws HttpException if the requested range cannot be satisfied.
+     * @throws RangeNotSatisfiableHttpException if the requested range cannot be satisfied.
```
Comment (dynasource) [inline framework/web/Response.php]: please make this line consistent with the other method at line 485

**Review by dynasource — CHANGES_REQUESTED** [review]

**Comment by samdark** [comment]
Merged with adjustments suggested. Test is modified to check that correct exception is thrown. Thanks for contribution.

**Pull request closed without merging**